toolkit/tools/internal/file/filecopybuilder.go

```diff
@@ -56,49 +55,98 @@ func (b FileCopyBuilder) Run() (err error) {
 	}
 
 	if b.NoDereference {
-		isSrcFileOrSymlink, err := IsFileOrSymlink(b.Src)
+		// Check if file is a symlink.
+		srcFileInfo, err := os.Lstat(b.Src)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to read source file link info:\n%w", err)
 		}
-		if !isSrcFileOrSymlink {
-			return fmt.Errorf("source (%s) is not a file or a symlink", b.Src)
-		}
-	} else {
-		isSrcFile, err := IsFile(b.Src)
-		if err != nil {
-			return err
+
+		isSrcSymlink := srcFileInfo.Mode().Type() == os.ModeSymlink
+		if isSrcSymlink {
+			// Copy the symlink.
+			symlinkPath, err := os.Readlink(b.Src)
+			if err != nil {
+				return fmt.Errorf("failed to read source symlink:\n%w", err)
+			}
+
+			err = os.Symlink(symlinkPath, b.Dst)
+			if err != nil {
+				return fmt.Errorf("failed to copy symlink:\n%w", err)
+			}
+
+			return nil
 		}
-		if !isSrcFile {
-			return fmt.Errorf("source (%s) is not a file", b.Src)
+	}
+
+	srcFileInfo, err := os.Stat(b.Src)
+	if err != nil {
+		return fmt.Errorf("failed to read source file info:\n%w", err)
+	}
+
+	isSrcFile := !srcFileInfo.IsDir()
+	if !isSrcFile {
+		return fmt.Errorf("source (%s) is not a file", b.Src)
+	}
+
+	// Open source file.
+	srcFile, err := os.Open(b.Src)
+	if err != nil {
+		return fmt.Errorf("failed to open source file:\n%w", err)
+	}
+	defer func() {
+		if srcFile != nil {
+			srcFile.Close()
 		}
+	}()
+
+	dstFileMode := b.FileMode
+	if !b.ChangeFileMode {
+		// Copy the source file's permissions.
+		dstFileMode = srcFileInfo.Mode().Perm()
 	}
 
 	err = CreateDestinationDir(b.Dst, b.DirFileMode)
 	if err != nil {
 		return fmt.Errorf("failed to create destination directory (%s):\n%w", b.Dst, err)
 	}
 
-	args := []string(nil)
-	if b.NoDereference {
-		args = append(args, "--no-dereference")
+	// Open/create destination file.
+	dstFile, err := os.OpenFile(b.Dst, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, dstFileMode)
+	if err != nil {
+		return fmt.Errorf("failed to create destination file:\n%w", err)
 	}
+	defer func() {
+		if dstFile != nil {
+			dstFile.Close()
+		}
+	}()
 
-	args = append(args, "--preserve=mode", b.Src, b.Dst)
+	// The permissions given to OpenFile is subject to umask.
+	// So, apply the permissions to ensure they match exactly.
+	err = dstFile.Chmod(dstFileMode)
```

Comment:
Chris, go's Chmod seems to drop the higher order bits from file mode.
See: https://github.com/microsoft/azure-linux-image-tools/blob/8e1841d911a662996c253a8b1171fb295125c1e6/toolkit/tools/internal/initrdutils/initrdutils.go#L131

Reply:
The implementation of Go's chmod function seems to have logic to handle setuid, setgid, and setsticky: https://github.com/golang/go/blob/master/src/os/file_posix.go#L60

And I tried using `additionalFiles` with files that had the setuid, setgid, and setsticky bits set and it worked.

So, I am not sure why that comment exists in `initrdutils.go`.

Reply:
thx Chris - I'll look more into why it didn't work for me.